cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.cpp

```diff
@@ -0,0 +1,272 @@
+        event.accept();
+      }
+    });
+  }
+
+  emit geoViewChanged();
+}
+
+void CoordinateConversionController::setCurrentPoint(const Point& p)
+{
+  if (p == m_currentPoint)
+    return;
+
+  m_currentPoint = p;
+  forceUpdateCoordinates();
+}
+
+void CoordinateConversionController::setCurrentPoint(
+  const QString& p,
+  CoordinateConversionOption* option)
+{
+  if (auto geoView = qobject_cast<GeoView*>(m_geoView))
+    setCurrentPoint(p, geoView->spatialReference(), option);
+  else
+    setCurrentPoint(p, SpatialReference(), option);
+}
+
+void CoordinateConversionController::setCurrentPoint(
+  const QString& p,
+  const SpatialReference& spatialReference,
+  CoordinateConversionOption* option)
+{
+  if (!option)
+    return;
+
+  auto point = option->pointFromString(p, spatialReference);
+  if (point.isValid())
+    setCurrentPoint(point);
+}
+
+QPointF CoordinateConversionController::screenCoordinate() const
+{
+  // TODO additional work required here to show a coordinate on the "edge" of
+  // the screen if coordinate is not in the current view.
+  QPointF res(-1.0, -1.0);
+  if (auto sceneView = qobject_cast<SceneView*>(m_geoView))
+  {
+    const auto location = sceneView->locationToScreen(m_currentPoint);
+    const auto lx = location.screenPoint().x();
+    const auto ly = location.screenPoint().y();
+
+    if (location.visibility() == SceneLocationVisibility::NotOnScreen)
+    {
+      // TODO attach to edge of screen.
+    }
+    else
+    {
+      res = QPointF(lx, ly);
+    }
+  }
+  else if (auto mapView = qobject_cast<MapView*>(m_geoView))
+  {
+    res = mapView->locationToScreen(m_currentPoint);
+  }
+
+  return res;
+}
+
+Point CoordinateConversionController::currentPoint() const
+{
+  return m_currentPoint;
+}
+
+GenericListModel* CoordinateConversionController::coordinateFormats() const
+{
+  return m_coordinateFormats;
+}
+
+GenericListModel* CoordinateConversionController::conversionResults() const
+{
+  return m_conversionResults;
+}
+
+void CoordinateConversionController::zoomToCurrentPoint()
+{
+  if (auto sceneView = qobject_cast<SceneView*>(m_geoView))
+  {
+    const Camera currentCam = sceneView->currentViewpointCamera();
+    const Camera newCam(m_currentPoint, m_zoomToDistance, currentCam.heading(), currentCam.pitch(), currentCam.roll());
+    sceneView->setViewpointCamera(newCam, 1.0);
+  }
+  else if (auto mapView = qobject_cast<MapView*>(m_geoView))
+  {
+    const Viewpoint currVP = mapView->currentViewpoint(ViewpointType::CenterAndScale);
+    const Viewpoint newViewPoint(m_currentPoint, currVP.targetScale());
+    mapView->setViewpoint(newViewPoint, 1.0);
+  }
+}
+
+void CoordinateConversionController::addNewCoordinateResultForOption(CoordinateConversionOption* option)
+{
+  auto result = new CoordinateConversionResult(m_conversionResults);
+  result->setType(option);
+  result->updateCoordinatePoint(currentPoint());
+  m_conversionResults->append(result);
```

Comment:
seems like the append function could be replaced with one which actually does the creation (e.g. take in the type and point)

Reply:
This is an interesting idea! A bit like `std::make_shared` and how it takes arbitrary arguments?